OpenAI Assistants v2 api support for OpenAIAssistantRunnable

codeonprod opened the pull request:
**Title**: "langchain: OpenAI Assistants v2 api support"

***Descriptions*** 
- [x] "attachments" support added along with backward compatibility of "file_ids"
- [x]  "tool_resources" support added while creating new assistant

- [ ] "tool_choice" parameter support
- [ ]  Streaming support


- **Dependencies:** OpenAI v2 API (openai>=1.23.0)
- **Twitter handle:** @skanta_rath

codeonprod committed:
Use attachments in OpenAIAssistantRunnable

In latest OpenAI Assistants API, 'file_ids' are not accepted and replaced by 'attachments'.
So, this PR proposes usage of the new 'attachments' by keeping the existing 'file_ids' intact and merging both the keys data.
The files provided by existing 'file_ids' will be turned on for file_search as well as code_interpreter by default.

Dependencies:
Python 3.11

langchain==0.1.17
langchain-openai==0.1.6
openai>=1.23.0

codeonprod committed:
1 occurrence of attachments skipped

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'master' into patch-1

mbonet commented:
Hello! @eyurtsev @Anush008 @efriis  @cbornet

Do you think you could help us here reviewing and approving? This is blocking the new OpenAI v2 Assistants and it's quite urgent :cry: 

Thank you very much! :heart:

cbornet commented:
Will older versions of openai <1.23.6 still work after this change ?

Enase commented:
mb unrelated question but OpenAI v2 Assistant now supports [tool_choice parameter](https://platform.openai.com/docs/api-reference/runs/object#runs/object-tool_choice) and [streaming](https://platform.openai.com/docs/assistants/overview/step-4-create-a-run?context=with-streaming). Does it covered?

Surya-Kanta commented:
> Will older versions of openai <1.23.6 still work after this change ?

Yes. They will work as these changes are backward compatible

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Added support for assistant tool_resources

Added tool_resources key support while creating assistants in OpenAIRssistantRunnable

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'langchain-ai:master' into patch-1

mbonet commented:
@Surya-Kanta thanks for your fast iteration! 

@Enase, Surya just added compatibility with tool_choice. As for streaming is not implemented yet. Please keep in mind that this PR is bugfixing not a new feature. Currently the assistants can't be migrated to v2 :cry:

Surya-Kanta commented:
@mbonet I will then continue working on complete migration of Assistants v2 in this branch.

mbonet commented:
@cbornet @Enase can we then merge? :pray:

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
OpenAIAssistantRunnable: remove attachments from create assistants

codeonprod committed:
OpenAIAssistantRunnable: tool_resources support added for thread create and run

codeonprod committed:
OpenAIAssistantRunnable: default openai model assistant v2 header set

Surya-Kanta commented:
I have already finished upgrade for all breaking changes related to v2 migration apart from streaming related changes. Anyone want to contribute in this PR are welcome in advance.

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'master' into patch-1

dentro-innovation commented:
would be great to have this in the next release, as it's required to use the Assistants API v2 when you wanna use it in combination with GPT4o

codeonprod committed:
Merge branch 'master' into patch-1

dentro-innovation commented on libs/langchain/langchain/agents/openai_assistant/base.py:
```diff
@@ -225,6 +243,7 @@ def create_assistant(
         name: str,
         instructions: str,
         tools: Sequence[Union[BaseTool, dict]],
+        tool_resources: Optional[Union[AssistantToolResources, dict]] | NotGiven = NOT_GIVEN,
```
did you actually try to use the code you changed in the PR yourself?

I just cloned your repo, and when trying to run the example from the docs I get an error.
Example code snippet:
```
from langchain.agents.openai_assistant import OpenAIAssistantRunnable

interpreter_assistant = OpenAIAssistantRunnable.create_assistant(
    name="langchain assistant",
    instructions="You are a personal math tutor. Write and run code to answer math questions.",
    model="gpt-4-turbo",
    tools=[{"type": "code_interpreter"}],
)
```

Error:
```
File [~/workingpath/langchain/libs/langchain/langchain/agents/openai_assistant/base.py:247](https://file+.vscode-resource.vscode-cdn.net/home/dachsteinhustler/magic/contract_work/choice/choicegpt/assistants_api_jupyter/~/magic/llm_apps/third_party_projects/langchain/libs/langchain/langchain/agents/openai_assistant/base.py:247)
    model: str,
    ^
SyntaxError: parameter without a default follows parameter with a default
```

So you need to move `tool_resources` below `model`

Enase commented:
> @Enase, Surya just added compatibility with tool_choice

It's not true even if you fix the obvious issues, thanks for your effort but I don't get the reason why you declare something that wasn't done.

codeonprod committed:
Merge branch 'master' into patch-1

codeonprod committed:
Merge branch 'master' into patch-1

ccurme committed:
update

ccurme committed:
lint

ccurme committed:
move to separate class

ccurme committed:
mark beta

ccurme commented:
Hi all, thank you for this. To avoid breaking changes with earlier versions of the openai sdk I've moved this to a new class `OpenAIAssistantV2Runnable`. I've marked it as beta. Please post your feedback; PRs are also welcome.

mbonet commented:
@ccurme thanks!

josantosc commented:
Hello, has the update been released?

ccurme commented:
> Hello, has the update been released?

This was moved to langchain-community and released today. If you `pip install langchain-community==0.2.1` you can import it with:
```python
from langchain_community.agents.openai_assistant import OpenAIAssistantV2Runnable
```

API reference: https://api.python.langchain.com/en/latest/agents/langchain_community.agents.openai_assistant.base.OpenAIAssistantV2Runnable.html

Note that this class is marked as `beta`; we welcome issues / PRs if you run into anything.

josantosc commented:
I identified that the function does not return the assistent tools, is there a way to make this change?

```python
        try:
            response = self._get_response(run)
```

dar747 commented:
This is great. Does it support streaming yet?

mbonet commented:
@ccurme do you think we can improve the original docs?  
https://api.python.langchain.com/en/latest/agents/langchain.agents.openai_assistant.base_v2.OpenAIAssistantV2Runnable.html

The suggested import is quite wrong `langchain.agents` no?
``` 
from langchain.agents.openai_assistant import OpenAIAssistantV2Runnable
```

You pointed it should be from `langchain_community` namespace and it took me a good 2 hrs to notice this discrepancy :cry:

DLOVRIC2 commented:
Looks like the pipeline is failing for this merge? Any indication when this will be successfully merged? 

@codeonprod thanks for doing this!
